### ISSUE TITLE
[Epic] Add more user-relevant information to explain plans

### ISSUE THREAD

**Issue by aalexandrov** [MaterializeInc/materialize#13138]
After #13137, we should focus on extending the text-based and graphical `EXPLAIN` outputs with more information that might be useful for understanding and self-managing Materialize clusters.  

### Goals

Make self-managing of Materialize clusters easier by surfacing:
1) clearer information.
2) more information.

### Acceptance Criteria

It is possible for a user with some experience parsing explanations generated by an RDBMS system to:
* have an idea of how a join works without having to consult documentation.
* determine how a one-off query is actually being rendered.
* determine whether the indexes they created are being used.
* figure out if the plan contains something bad, such a cross join.

There exists a framework to allow the future addition of relevant factors in the process to decide on a plan.

### Out of scope

* Retaining the original column names in plans. Our IRs at the moment do not contain the column identifiers required to track what columns were originally named, and significant refactoring from the top down would be required to add them.
* Tracking which parts of a lower IR correspond to a node in a higher level IR (1st proposal in this comment). 
* Improving dataflow graph visualization, which is technically a 

### Tasks

- [x] #13482
- [x] Limit the max number of items that would be printed as part of a constant.
- [x] Print a join on empty keys as "Cross join".
- [ ] #13485
- [x] #13484
- [x] #12629
- [x] Show whether or not a one-off query is using fast or slow path.
- [x] https://github.com/MaterializeInc/materialize/issues/4887
- [x] Design a framework for exposing additional information.
- [x] #14315
- [ ] #14842
- [x] Take into account the fact that different sets of plan annotations / additional information might be useful for different purposes.
     - We have added the ability to disable seeing whether or not a fast path is used, the ability to have more detailed vs. less detailed join plans.
- [x] Add documentation (incl. physical plan + unified format).

### Testing

TODO

### Bugs

TODO

### Related

- https://github.com/MaterializeInc/materialize/issues/13137
- https://github.com/MaterializeInc/product/issues/17
- https://github.com/MaterializeInc/materialize/issues/6201
- https://github.com/MaterializeInc/materialize/pull/12597
- https://github.com/MaterializeInc/materialize/issues/7301
- https://materializeinc.slack.com/archives/C02CB7L4TCG/p1652926444374379
- https://www.notion.so/materialize/What-EXPLAIN-semantics-do-we-want-a17b6786dc1c4239aa9ebbf8fae9cdaa

### Questions

- Should we change when the explain stages happen and how many of them there are?

**Comment by uce** [comment]
@aalexandrov What's the plan with the remaining subtasks? No need to pick this up. I'm trying to determine whether we want to still work on them or close this epic as is.

**Comment by aalexandrov** [comment]
There is one remaining subtask that can be resolved separately.
